// @mdxeditor/editor/dist/styles/lexical-theme.module.css.js
const light = "_light_1boox_1";
const lightTheme = "_light-theme_1boox_1";
const dark = "_dark_1boox_1";
const darkTheme = "_dark-theme_1boox_1";
const bold = "_bold_1boox_10";
const italic = "_italic_1boox_14";
const underline = "_underline_1boox_18";
const strikethrough = "_strikethrough_1boox_34";
const underlineStrikethrough = "_underlineStrikethrough_1boox_38";
const subscript = "_subscript_1boox_42";
const superscript = "_superscript_1boox_47";
const code = "_code_1boox_52";
const nestedListItem = "_nestedListItem_1boox_59";
const listitem = "_listitem_1boox_69";
const listItemChecked = "_listItemChecked_1boox_73";
const listItemUnchecked = "_listItemUnchecked_1boox_74";
const admonitionDanger = "_admonitionDanger_1boox_151";
const admonitionInfo = "_admonitionInfo_1boox_152";
const admonitionNote = "_admonitionNote_1boox_153";
const admonitionTip = "_admonitionTip_1boox_154";
const admonitionCaution = "_admonitionCaution_1boox_155";
const mdxTextExpression = "_mdxTextExpression_1boox_188";
const lexicalThemeStyles = {
  light,
  lightTheme,
  dark,
  darkTheme,
  bold,
  italic,
  underline,
  strikethrough,
  underlineStrikethrough,
  subscript,
  superscript,
  code,
  nestedListItem,
  listitem,
  listItemChecked,
  listItemUnchecked,
  admonitionDanger,
  admonitionInfo,
  admonitionNote,
  admonitionTip,
  admonitionCaution,
  mdxTextExpression
};
export {
  admonitionCaution,
  admonitionDanger,
  admonitionInfo,
  admonitionNote,
  admonitionTip,
  bold,
  code,
  dark,
  darkTheme,
  lexicalThemeStyles as default,
  italic,
  light,
  lightTheme,
  listItemChecked,
  listItemUnchecked,
  listitem,
  mdxTextExpression,
  nestedListItem,
  strikethrough,
  subscript,
  superscript,
  underline,
  underlineStrikethrough
};
